nav2_dynamic_params/include/nav2_dynamic_params/dynamic_params_client.hpp

```diff
@@ -230,26 +195,106 @@ class DynamicParamsClient
     return get_event_param_or<T>(
       node_->get_namespace(), node_->get_name(), param_name, new_value, default_value);
   }
-
   // A check to filter whether parameter name is part of the lastest event
-  bool is_in_event(const std::string & name)
+  bool is_in_event(const std::string & path, const std::string & param_name)
   {
-    rclcpp::ParameterEventsFilter filter(last_event_, {name},
+    auto full_path = path;
+    if (*full_path.begin() != '/') {
+      full_path = '/' + full_path;
+    }
+
+    rclcpp::ParameterEventsFilter filter(last_event_, {param_name},
       {rclcpp::ParameterEventsFilter::EventType::NEW,
         rclcpp::ParameterEventsFilter::EventType::CHANGED});
-    return !filter.get_events().empty();
+
+    return full_path == last_event_->node && !filter.get_events().empty();
   }
-  // Variant of is_in_event to also check the node path of last event
-  bool is_in_event(const std::string & path, const std::string & name)
+
+  // Variant of is_in_event to specify namespace and node name
+  bool is_in_event(
+    const std::string & name_space,
+    const std::string & node_name, const std::string & param_name)
   {
-    auto full_path = path;
-    if (*full_path.begin() != '/') {
-      full_path = '/' + full_path;
+    auto full_path = join_path(name_space, node_name);
+    return is_in_event(full_path, param_name);
+  }
+
+  // Variant of is_in_event to check under member node
+  bool is_in_event(const std::string & param_name)
+  {
+    return is_in_event(node_->get_namespace(), node_->get_name(), param_name);
+  }
+
+protected:
+  void event_callback(const rcl_interfaces::msg::ParameterEvent::SharedPtr event)
+  {
+    last_event_ = event;
+    if (is_event_in_map(event)) {
+      user_callback_();
     }
-    return full_path == last_event_->node && is_in_event(name);
   }
 
 private:
+  std::vector<rclcpp::Parameter> get_params_from_node(
+    const std::string & path,
+    const std::vector<std::string> & param_names, int attempts_max = 5)
+  {
+    std::vector<rclcpp::Parameter> params;
+    if (path == join_path(node_->get_namespace(), node_->get_name())) {
+      if (param_names.size() < 1) {
+        auto param_list = node_->list_parameters({}, 1);
+        params = node_->get_parameters(param_list.names);
+      } else {
+        params = node_->get_parameters(param_names);
+      }
+      return params;
+    } else {
+      bool success = false;
+      int attempts = 0;
+      while (!success && attempts < attempts_max) {
+        auto client = std::make_shared<rclcpp::AsyncParametersClient>(node_, path);
+        if (param_names.size() < 1) {
+          auto param_list_future = client->list_parameters({}, 1);
+          if (rclcpp::spin_until_future_complete(
```

Comment:
this same call to spin is checked on three places, maybe make a separate function